Right-clicking on reagent containers picks previous reagent transfer amount

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Reopening #59351

- Right-clicking a reagent container in your active hand will pick the previous transfer amount instead of the next one
- Adds an attack_self_secondary proc which allows for different interactions when right-clicking things in your active hand

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Should solve the tedium of picking reagent transfer amounts. You don't have to click as many times to pick the amount you want and if you overshoot you don't have to go through the entire list again.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: cacogen
qol: Right-clicking a reagent container in your active hand will pick the previous transfer amount instead of the next
code: Mauna mug uses ctrl+click instead of attack_self() to toggle power
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

cacogen committed:
Right-click picks previous reagent transfer amount and attack_self_secondary

- Right-clicking a reagent container in your active hand will pick the previous transfer amount
- Adds an attack_self_secondary proc which allows for different interactions when right-clicking things in your active hand
- Not sure if I'll add a matching hotkey and verb or not

cacogen committed:
Stylises balloon_alert to all lowercase

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

cacogen committed:
Removes unnecessary duplication of interact() on in-hand right-click

cacogen committed:
Apply suggestions from code review

cacogen committed:
Indent

using code review suggestions to edit my code while working on something else is not the most optimal way of contributing

cacogen committed:
Makes attack_hand_secondary work with spray bottles

cacogen committed:
Is this what he meant by autodoc

cacogen committed:
Solves issues with reagent_containers that override attack_self

cacogen committed:
Intents

cacogen commented:
@Mothblocks I went through every reagent_container that overrides attack_self and made them work with this. Hopefully there won't be any issues with things being able to change transfer amount or spray distance when they shouldn't. Wondering what you think though.

Rohesie commented on code/modules/reagents/reagent_containers/spray.dm:
```diff
@@ -132,14 +133,22 @@
 	qdel(reagent_puff)
 
 /obj/item/reagent_containers/spray/attack_self(mob/user)
+	..()
+	toggle_stream_mode(user)
+
+/obj/item/reagent_containers/spray/attack_self_secondary(mob/user)
+	..()
```
```suggestion
	. = ..()
	toggle_stream_mode(user)

/obj/item/reagent_containers/spray/attack_self_secondary(mob/user)
	. = ..()
```
* Preserve parent call return by default, unless if you have a good reason not to.

cacogen replied:
I don't even know what this does lol

I get that . is what we return with because I see and add stuff like `. += "<span class='notice'>Ctrl+click to toggle power.</span>"` under `examine()` where it's adding to a list to eventually format and send to the player

And I think that if you went `return ..()` here you'd stop the rest of the code executing

But what is the difference between `. = ..()` and `..()`? Does one call the parent proc immediately and the other wait until the rest of the code is executed or something?

Rohesie replied:
It preserves the parent call return.
`..()` will call the parent proc, which will return some kind of value, `null` if nothing else.
That value will be lost if not preserved somewhere, and that's that the dot variable is doing: passing it forward.

We try to orient our code so that parent returns trickle down the line, and redefinitions that call parent are mindful of it.

cacogen committed:
dot equals dot dot empty breast

Co-authored-by: Rohesie <rohesie@gmail.com>

Rohesie commented on code/modules/reagents/reagent_containers.dm:
```diff
@@ -47,17 +53,35 @@
+/obj/item/reagent_containers/proc/mode_change_message(mob/user)
+	return
+
+/obj/item/reagent_containers/proc/change_transfer_amount(mob/user, direction)
+	if(!possible_transfer_amounts.len)
+		return
+	var/list_position = 0
+	for(var/transfer_amount in possible_transfer_amounts)
+		list_position++
+		if(transfer_amount == amount_per_transfer_from_this)
+			switch(direction)
+				if(FORWARD)
+					if(list_position < possible_transfer_amounts.len) //not at end
+						amount_per_transfer_from_this = possible_transfer_amounts[list_position + 1] //move to next entry
+					else
+						amount_per_transfer_from_this = possible_transfer_amounts[1] //move to list start
+				if(BACKWARD)
+					if(list_position == 1) //at list start
+						amount_per_transfer_from_this = possible_transfer_amounts[possible_transfer_amounts.len] //move to list end
+					else
+						amount_per_transfer_from_this = possible_transfer_amounts[list_position - 1] //move to previous entry
+			balloon_alert(user, "transferring [amount_per_transfer_from_this]u")
+			mode_change_message(user)
+			return
```
I'd keep the current list_position as an object variable, personally, instead of searching for the position each time.
Tne you can simply do this:


```suggestion
/obj/item/reagent_containers/proc/change_transfer_amount(mob/user, direction)
	var/list_len = length(possible_transfer_amounts)
	if(!list_len)
		return
	switch(direction)
		if(FORWARD)
			list_position = (list_position % list_len) + 1
		if(BACKWARD)
			list_position = (list_position - 1) || list_len
	amount_per_transfer_from_this = possible_transfer_amounts[list_position]
	balloon_alert(user, "transferring [amount_per_transfer_from_this]u")
	mode_change_message(user)
```
The modulo operator will result in `0` if the `list_position` value matches the length, and thus increment it back to `1`, looping from last to first position.
The `||` operator will substitute the value if falsey (as `0`) for the one on the right side, this being the length of the list, thus looping from first to last position.

I'd also turn `possible_transfer_amounts` into a lazy list as well, but that may be out of scope.

cacogen replied:
Thank you, this is clever

cacogen committed:
Utilises the power of mathematics

Co-authored-by: Rohesie <rohesie@gmail.com>

Rohesie commented:
This implies you save on the object the current list position. Ideally starting with 1 and with a better variable name.

cacogen commented:
Yeah, I added the object variable and tested it locally. I'll think of a better name, though.

cacogen committed:
Adds amount_list_position var

- On reagent containers with one transfer amount that use amount_per_transfer_from_this, adds that transfer amount to their possible_transfer_amounts list, so attempting to change transfer amount will show the current transfer amount instead (as far as I'm aware there's no other way to see the transfer amount on reagent containers that can't change it)
- Adds second examine message specific to reagent containers with only one transfer amount
